easypost/resource.py

```diff
@@ -0,0 +1,284 @@
+from easypost.error import Error
+
+
+class Resource(EasyPostObject):
+    def _ident(self):
+        return [self.get("id")]
+
+    @classmethod
+    def retrieve(cls, easypost_id, api_key=None, **params):
+        try:
+            easypost_id = easypost_id["id"]
+        except (KeyError, TypeError):
+            pass
+
+        instance = cls(easypost_id, api_key, **params)
+        instance.refresh()
+        return instance
+
+    def refresh(self):
+        requestor = Requestor(self._api_key)
+        url = self.instance_url()
+        response, api_key = requestor.request("get", url, self._retrieve_params)
+        self.refresh_from(response, api_key)
+        return self
+
+    @classmethod
+    def snakecase_name(cls):
+        """Return the class name as snake_case."""
+        snake_name = (cls.__name__[0:1] + re.sub(r"([A-Z])", r"_\1", cls.__name__[1:])).lower()
+        return snake_name
+
+    @classmethod
+    def class_url(cls):
+        cls_name = cls.snakecase_name()
+        if cls_name[-1:] in ("s", "h"):
+            return "/%ses" % cls_name
+        else:
+            return "/%ss" % cls_name
+
+    def instance_url(self):
+        easypost_id = self.get("id")
+        if not easypost_id:
+            raise Error("%s instance has invalid ID: %r" % (type(self).__name__, easypost_id))
+        return "%s/%s" % (self.class_url(), easypost_id)
+
+
+class AllResource(Resource):
+    @classmethod
+    def all(cls, api_key=None, **params):
+        requestor = Requestor(api_key)
+        url = cls.class_url()
+        response, api_key = requestor.request("get", url, params)
+        return util.convert_to_easypost_object(response, api_key)
+
+
+class CreateResource(Resource):
+    @classmethod
+    def create(cls, api_key=None, **params):
+        requestor = Requestor(api_key)
+        url = cls.class_url()
+        wrapped_params = {cls.snakecase_name(): params}
+        response, api_key = requestor.request("post", url, wrapped_params)
+        return util.convert_to_easypost_object(response, api_key)
+
+
+class UpdateResource(Resource):
+    def save(self):
+        if self._unsaved_values:
+            requestor = Requestor(self._api_key)
+            params = {}
+            for k in self._unsaved_values:
+                params[k] = getattr(self, k)
+                if type(params[k]) is EasyPostObject:
+                    params[k] = params[k].flatten_unsaved()
+            params = {self.snakecase_name(): params}
+            url = self.instance_url()
+            response, api_key = requestor.request("put", url, params)
+            self.refresh_from(response, api_key)
+
+        return self
+
+
+class DeleteResource(Resource):
+    def delete(self, **params):
+        requestor = Requestor(self._api_key)
+        url = self.instance_url()
+        response, api_key = requestor.request("delete", url, params)
+        self.refresh_from(response, api_key)
+        return self
+
+
+class Requestor(object):
```

Comment:
I'm still unsure how I feel about this class living in this module... will need to think on it.

Reply:
We were able to break this back out to its own module.